multinet/api/utils/arango.py

```diff
@@ -68,6 +76,28 @@ def from_collections(db: StandardDatabase, collections: List[str]) -> ArangoQuer
 
         return ArangoQuery(db, query_str=query_str, bind_vars=bind_vars)
 
+    def filter(self, doc: Dict) -> ArangoQuery:
+        """Filter an AQL query to match the provided document."""
+        # If empty filter, do nothing
+        if not doc:
+            return self
+
+        new_bind_vars = dict(self.bind_vars)
+        filter_query_lines = []
+        for k, v in doc.items():
+            # Create unique bind var keys, so they don't
+            # conflict with any previous or future keys
+            key_key = str(uuid.uuid4())[:8]
+            val_key = str(uuid.uuid4())[:8]
+
+            new_bind_vars[key_key] = k
+            new_bind_vars[val_key] = v
+            filter_query_lines.append(f'doc[@{key_key}] == @{val_key}')
```

Comment:
Can you help me understand what this is doing (especially with bind vars)?

Reply:
Sure, this is a bit confusing.

To filter an existing AQL query with a document, we need to translate the following format
```json
{ "foo": "bar" }
```
into
```
FILTER doc['foo'] == 'bar'
```

Each `ArangoQuery` instance has a `bind_vars` member, which is used to store any variables in the request (as is [recommended](https://www.arangodb.com/docs/stable/aql/fundamentals-bind-parameters.html) by the Arango docs). Since we have several variables here (each key and each value), we want to include those in `bind_vars`. However, for each variable we include, we need to name that variable something, and reference that name in the query itself. This section of code creates a unique key for each variable used, and then references that unique key in the query itself. So in regards to the above example document, it would produce something like the following:

```
# bind_vars
{
  "f2e59b48": "foo",
  "a7f3755d": "bar"
}

# in aql, @ is needed to reference bind variables
FILTER doc[@f2e59b48] == @a7f3755d

```

Hopefully that makes sense.

Reply:
Yes, that does make sense, thanks!

Two things:
1. I think a code comment explaining what's going on here would be good to include.
2. If the UUID begins with a number, will that cause Arango any trouble?

Reply:
1. Will do
2. Good question. From the arango docs:
> The syntax for bind parameters is `@name` where `@` signifies that this is a bind parameter and name is the actual parameter name. Parameter names must start with any of the letters a to z (upper or lower case) or a digit (0 to 9), and can be followed by any letter, digit or the underscore symbol.

So it looks like we're good there.

Reply:
50cd001

Reply:
This is good, though I meant something more like, provide a bit more detail or clarity about what this function is actually doing in its docstring (i.e., I was asking for an improved high-level description of the function).

Reply:
Sure, I'll add that in.